dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/CategoryOptionComboMapGenerator.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2004-2022, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.predictor;
+
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toMap;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.hisp.dhis.category.CategoryCombo;
+import org.hisp.dhis.category.CategoryOption;
+import org.hisp.dhis.category.CategoryOptionCombo;
+import org.hisp.dhis.common.MapMap;
+
+/**
+ * Generator of a map that, for each category combo (CC) found in an input data
+ * element, returns a map from the category option combinations of that CC to
+ * the category option combinations of the output category combo.
+ *
+ * @author Jim Grace
+ */
+public class CategoryOptionComboMapGenerator
+{
+    private CategoryOptionComboMapGenerator()
+    {
+    }
+
+    /**
+     * Returns a map that, for each input category combo (CC), returns a map
+     * from the category option combinations of that CC to the category option
+     * combinations of the output category combo.
+     * <p>
+     * The map has entries only for those CCs where all the COCs can be mapped
+     * to output COCs.
+     *
+     * @param outputCombo predictor output data element category combination.
+     * @param inputCategoryCombos data element cat combos found in the input.
+     * @return the map.
+     */
+    public static MapMap<String, String, String> getCcMap( CategoryCombo outputCombo,
```

Comment:
The `MapMap` approach has the drawback of being a poor way of communicating what the data represents or how it is organized. I think with lombok allowing us to define data structures with just a class with some fields and some class level annotations it often a better option to introduce a new "record" type. That way the semantics get encoded.

Reply:
It surely is easier to grasp now. Extending collection or map classes comes with its own issues but as long as this is used in a small scope those aren't that bad. Something that would be better than maps of maps is if this can be modelled as list of records. If this problem is not suited for that either extending a map or creating a record with a map field is fine. I personally would rather have a map field but here I think extension is fine too.

Reply:
@jbee Understood. Ultimately a Map is needed so that lookups can happen quickly as data (potentially lots of it) is processed. The MapMap (or Map of Map) is nice because two different decisions are needed as the data is processed:
1. Should the COCs of a CategoryCombo (CC) be translated? Easy, see if the top map has that key.
2. To what output COC should an input COC (off a CC) be translated? Access both levels of map.

For the second use case, a single level of Map could be used with the key being perhaps a concatenation of the CC UID with the COC UID. Then for the first use case a Set might be used to contain the set of CCs that have map entries (combined with COCs) in the Map.

It's perhaps a tradeoff being complexity and more code vs. being more transparent. A judgement call to be sure. More code in the sake of greater transparency can also increase the time to understand and modify the code correctly which is, in the end, what we are after.

If you think the code now is reasonable enough, please re-approve the PR. (I had re-requested a review from you.) I need to move on to other things between now and the Annual Conference. :)

I appreciate the time and care you take in reviewing my PRs. I believe I am improving my craft as a result!

Reply:
I missed that you re-requested but you had my approval ;)